docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,388 @@
+force deleting a pod means that the pod processes may continue
+to run for an arbitary amount of time. If a higher level component like the
+PetSet controller treats the existence of the pod API object as a strongly
+consistent entity, deleting the pod in this fashion will violate the
+at-most-one guarantees we wish to offer for pet sets.
+
+
+### Guarantees provided by replica sets and replication controllers
+
+ReplicaSets and ReplicationControllers both attempt to preserve availability
+of their constituent pods over ensuring **at most one** semantics. So a
+replica set to scale 1 will immediately create a new pod when it observes an
+old pod is delete, and as a result at many points in the lifetime of a replica
+set there will be 2 copies of a pod's processes running concurrently. Only
+access to exclusive resources like storage can prevent that simultaneous
+execution.
+
+Deployments, being based on replica sets, can offer no stronger guarantee.
+
+
+### Concurrent access guarantees for shared storage
+
+A persistent volume that references a strongly consistent storage backend
+like AWS EBS, GCE PD, OpenStack Cinder, or Ceph RBD can rely on the storage
+API to prevent corruption of the data due to simultaneous access by multiple
+clients. However, many commonly deployed storage technologies in the
+enterprise offer no such consistency guarantee, or much weaker variants, and
+rely on complex systems to control which clients may access the storage.
+
+If a PV is assigned a iSCSI, Fibre Channel, or NFS mount point and that PV
+is used by two pods on different nodes simultaneously, concurrent access may
+result in corruption, even if the PV or PVC is identified as "read write one".
+PVC consumers must ensure these volume types are *never* referenced from
+mulitple pods without some external synchronization. As described above, it
+is not safe to use persistent volumes that lack RWO guarantees with a
+replica set or deployment, even at scale 1.
+
+
+## Proposed changes
+
+### Avoid multiple instances of pods
+
+To ensure that the Pet Set controller can safely use pods and ensure at most
+one pod instance is running on the cluster at any time for a given pod name,
+it must be possible to make pod deletion strongly consistent.
+
+To do that, we will:
+
+* Give the Kubelet sole responsibility for normal deletion of pods -
+  only the Kubelet in the course of normal operation should ever remove a
+  pod from etcd (only the Kubelet should force delete)
+  * The kubelet must not delete the pod until all processes are confirmed
+    terminated.
+* Application owners must be free to force delete pods, but they *must*
+  understand the implications of doing so, and all client UI must be able
+  to communicate those implications.
+* All existing controllers in the system must be limited signaling pod
+  termination (starting graceful deletion), and are not allowed to force
+  delete a pod.
```

Comment:
Yeah, and forgiveness is a delay before a delete *that is triggered by placement of a taint* so there are plenty of deletion scenarios that forgiveness wouldn't help with. (It would help with node unreachable (Unknown node condition), though.)